src/ripple/app/tx/impl/AMMDeposit.cpp

```diff
@@ -0,0 +1,528 @@
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid Asset2InAmount";
+        return *res;
+    }
+
+    if (auto const res = invalidAmount(ePrice))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid EPrice";
+        return *res;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+AMMDeposit::preclaim(PreclaimContext const& ctx)
+{
+    auto const accountID = ctx.tx[sfAccount];
+
+    if (!ctx.view.read(keylet::account(accountID)))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: Invalid account.";
+        return terNO_ACCOUNT;
+    }
+
+    auto const ammSle = getAMMSle(ctx.view, ctx.tx[sfAMMID]);
+    if (!ammSle)
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: Invalid AMMID.";
+        return terNO_ACCOUNT;
+    }
+
+    auto const asset1In = ctx.tx[~sfAsset1In];
+    auto const asset2Out = ctx.tx[~sfAsset2Out];
+
+    if (asset1In)
+    {
+        if (auto const ter =
+                requireAuth(ctx.view, asset1In->issue(), accountID);
+            ter != tesSUCCESS)
+        {
+            JLOG(ctx.j.debug()) << "AMM Deposit: account is not authorized, "
+                                << asset1In->issue();
+            return ter;
+        }
+    }
+
+    if (asset2Out)
+    {
+        if (auto const ter =
+                requireAuth(ctx.view, asset2Out->issue(), accountID);
+            ter != tesSUCCESS)
+        {
+            JLOG(ctx.j.debug()) << "AMM Deposit: account is not authorized, "
+                                << asset2Out->issue();
+            return ter;
+        }
+    }
+
+    if (isFrozen(ctx.view, asset1In) || isFrozen(ctx.view, asset2Out))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit involves frozen asset.";
+        return tecFROZEN;
+    }
+
+    auto const [asset1, asset2, lptAMMBalance] =
+        ammHolds(ctx.view, *ammSle, std::nullopt, std::nullopt, ctx.j);
+    if (asset1 <= beast::zero || asset2 <= beast::zero ||
+        lptAMMBalance <= beast::zero)
+    {
+        JLOG(ctx.j.debug())
+            << "AMM Deposit: reserves or tokens balance is zero.";
+        return tecAMM_BALANCE;
+    }
+
+    if (auto const lpTokens = ctx.tx[~sfLPToken];
+        lpTokens && lpTokens->issue() != lptAMMBalance.issue())
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid LPTokens.";
+        return temBAD_AMM_TOKENS;
+    }
+
+    return tesSUCCESS;
+}
+
+void
+AMMDeposit::preCompute()
+{
+    return Transactor::preCompute();
+}
+
+std::pair<TER, bool>
+AMMDeposit::applyGuts(Sandbox& sb)
+{
+    auto const asset1In = ctx_.tx[~sfAsset1In];
+    auto const asset2In = ctx_.tx[~sfAsset2In];
+    auto const ePrice = ctx_.tx[~sfEPrice];
+    auto const lpTokensDeposit = ctx_.tx[~sfLPToken];
+    auto ammSle = getAMMSle(sb, ctx_.tx[sfAMMID]);
+    assert(ammSle);
+    auto const ammAccountID = ammSle->getAccountID(sfAMMAccount);
+
+    auto const tfee = getTradingFee(*ammSle, account_);
+
+    auto const [asset1, asset2, lptAMMBalance] = ammHolds(
```

Comment:
Here's an example of a function that can throw, but isn't wrapped in a try/catch block. Can the exception ever be triggered from here? If so, this needs to be wrapped in a try/catch and the appropriate error code returned from the function. If it doesn't need to be wrapped, we should add a comment explaining why.